drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+
+```ts
+@Component({
+  selector: "app-root",
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Full name: {{ fullName() }}</p>
+    <p>Age: {{ age() }}</p>
+  `,
+})
+export class AppComponent {
+  private activatedRoute = inject(ActivatedRoute);
+
+  firstName = this.getQueryParam("firstName");
+  lastName = this.getQueryParam("lastName");
+  age = this.getQueryParam("age");
+
+  fullName = computed(() => `${this.firstName() ?? ""} ${this.lastName() ?? ""}`);
+
+  private getQueryParam(name: string) {
+    const value$ = this.activatedRoute.queryParams.pipe(
+      map((allQueryParams) => allQueryParams[name])
+    );
+
+    return toSignal(value$);
+  }
+}
+```
+
+![Consuming query params with signals](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536530242/EJ6p_x7xp.gif?auto=format)
+
+This is already pretty cool! We have a handy utility to consume query params. Notice how easily we combined the `firstName` and `lastName` signals to compute the `fullName` one, this would have been much more verbose and complicated using `RxJS`.
+
+Let's add more features. Let's say we want a button to anonymise the user, meaning removing the values of `firstName` and `lastName`:
+
+```ts
+@Component({
+  selector: "app-root",
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Full name: {{ fullName() }}</p>
+    <p>Age: {{ age() }}</p>
+
+    <button (click)="anonymise()">I want to be anonymous 👤</button>
+  `,
+})
+export class AppComponent {
+  private activatedRoute = inject(ActivatedRoute);
+  private router = inject(Router);
+
+  // ...
+
+  anonymise() {
+    const queryParams = { firstName: undefined, lastName: undefined };
+    this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
+  }
+
+  // ...
+}
+```
+
+![Changing query params with Angular router](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536580024/nqZ9dSNZx.gif?auto=format)
+
+That's amazing! everything is working as expected, our signals are correctly getting the new `undefined` values and Angular is updating the UI accordingly.
+
+Let's take a closer look in this `anonymise` method: We are using Angular's router to navigate to the exact same page, that's the reason we are passing an empty array:
+
+```ts
+this.router.navigate([]);
+```
+
+We do this because we are only interested to change the values of the query params, not the route or anything else.
+
+In this case we want to unset the `firstName` and `lastName` query params, so we can anonymise the user. That's why we are passing `undefined` as their new values:
+
+```ts
+const queryParams = { firstName: undefined, lastName: undefined };
+this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
+```
+
+The `queryParamsHandling: "merge"` option is basically telling Angular to perform an **update** of the query params. This is important because, in this case, we are not interested to lose the `age` query param whenever it's there.
```

Comment:
**thought:** I think the whole section from line 154 to here could be reduced. Changing query params in Angular is not a remarkably unique or complicated task. Thinking about the target audience of this article, I would assume them to at least understand the code above and a very short summary (without repeating the snippets) would help everyone else enough.